serverless.yaml

```diff
@@ -1,362 +1,418 @@
-      ElasticSearchDomainKibanaEndpoint:
-        Condition: isDev
-        Description: ElasticSearch Kibana endpoint
-        Value: !Join ['', ['https://', !GetAtt ElasticSearchDomain.DomainEndpoint, '/_plugin/kibana']]
-      ElasticSearchKibanaUserPoolId:
-        Condition: isDev
-        Description: User pool id for the provisioning ES Kibana users.
-        Value: !Ref KibanaUserPool
-      ElasticSearchKibanaUserPoolAppClientId:
-        Condition: isDev
-        Description: App client id for the provisioning ES Kibana users.
-        Value: !Ref KibanaUserPoolClient
-      CloudwatchExecutionLogGroup:
-        Description: Cloudwatch Execution log group for storing request/responses for auditing purposes
-        Value: !Join ['', ['API-Gateway-Execution-Logs_', !Ref ApiGatewayRestApi, '/', '${self:custom.stage}']]
-        Export:
-          Name: CloudwatchExecutionLogGroup-${self:custom.stage}
-      CloudwatchExecutionLogGroupArn:
-        Description: Arn of Cloudwatch Execution log group for storing request/responses for auditing purposes
-        Value:
-          !Join [
-            '',
-            [
-              !Sub 'arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:API-Gateway-Execution-Logs_',
-              !Ref ApiGatewayRestApi,
-              '/',
-              '${self:custom.stage}',
-              ':log-stream:*',
-            ],
-          ]
-        Export:
-          Name: CloudwatchExecutionLogGroup-${self:custom.stage}-Arn
-
-plugins:
-  - serverless-plugin-typescript
-  - serverless-offline
+#
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  SPDX-License-Identifier: Apache-2.0
+#
+
+service: fhir-service
+
+custom:
+  oldResourceTableName: 'resource-${self:custom.stage}'
+  resourceTableName: 'resource-db-${self:custom.stage}'
+  stage: ${opt:stage, self:provider.stage}
+  region: ${opt:region, self:provider.region}
+  oauthCallback: ${opt:oauthCallback, 'http://localhost'}
+  oauthRedirect: ${opt:oauthRedirect, 'http://localhost'}
+  config: ${file(serverless_config.json)}
+
+provider:
+  name: aws
+  region: us-west-2
+  stage: dev
+  runtime: nodejs12.x
+  memorySize: 512
+  stackTags:
+    FHIR_SERVICE: !Sub 'fhir-service-${AWS::Region}-${self:custom.stage}'
+  environment:
+    API_URL:
+      !Join [
+        '',
+        ['https://', !Ref ApiGatewayRestApi, !Sub '.execute-api.${AWS::Region}.amazonaws.com/', !Ref Stage],
+      ]
+    S3_KMS_KEY:
+      Ref: S3KMSKey
+    RESOURCE_TABLE: '${self:custom.resourceTableName}'
+    FHIR_BINARY_BUCKET:
+      Ref: FHIRBinaryBucket
+    ELASTICSEARCH_DOMAIN_ENDPOINT: !Join ['', ['https://', !GetAtt ElasticSearchDomain.DomainEndpoint]]
+    OAUTH2_DOMAIN_ENDPOINT:
+      !Join ['', ['https://', !Ref UserPoolDomain, !Sub '.auth.${AWS::Region}.amazoncognito.com/oauth2']]
+  apiKeys:
+    - name: 'developer-key-${self:custom.stage}' # Full name must be known at package-time
```

Comment:
Do you want to use !Ref Stage here as well?

Reply:
`sls package` creates a Usage Plan for the API Key. Using !Ref Stage, doesn't work because Stage is not known to `sls package`, only variable (not Parameter) values are replaced when `sls package` is run. Doing !Ref Stage prevents `sls package` from completing with this error:

```
...
Serverless: Packaging service...
Serverless: Excluding development dependencies...

  Serverless Error ---------------------------------------

  API key "name" has no usage plan defined

  Get Support --------------------------------------------
...
```